### ISSUE TITLE
Question: What is this error?!?!

### ISSUE THREAD

**Issue by sizeak** [bit-team/backintime#931]
Could someone shed some light on what this error is about, I'm trying to run it as root.

net.launchpad.backintime.InvalidCmd: Parameter 'cmd' contains non-whitelisted cmd/parameter (/bin/nice)

Thanks!

**Comment by sizeak** [comment]
I've found that this error only occurs when I try and configure it to backup automatically when the drive is plugged in. If I use a fixed interval or on boot instead, I don't get an error.

**Comment by bachy** [comment]
i have the exact same error on arch, how can i help to fix this ?

**Comment by mcku** [comment]
try disabling "nice" options in the options tab

**Comment by brummer-simon** [comment]
I have the same Issue and disabling "nice" doesn't help because the next the invalid command complains is 
"/usr/sbin/backintime". I'll resort to creating backups based on an interval. If you need move details on the issue feel free to contact me.

**Comment by wgranicki** [comment]
> I have the same Issue and disabling "nice" doesn't help because the next the invalid command complains is "/usr/sbin/backintime". I'll resort to creating backups based on an interval. If you need move details on the issue feel free to contact me.

Same problem here

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Thanks for reporting the issue.

Please give detailed information's about
 - operating system: name, version, plattform (amd64, arm, ...)
 - BackInTime version
 - How was BackInTime installed

If possible please give us the full debug output (use `--debug` option when starting BiT).

Tell us the steps to reproduce that error.

Thanks in advance.

**Comment by wgranicki** [comment]
My setup:

- Manjaro Linux 21.3.7, kernel version 5.15.60, running on Intel i7-6820HK
- BackInTime version 1.3.2 installed from official Manjaro repository

Steps to reproduce:

- Start BackInTime
- Open Settings
- Change schedule to after disk connection (udev)
- Click OK
- Error message appears

Log (using `--debug` option):
https://pastebin.com/mErzUr0R

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Thanks for the information's. This is related to D-Bus which I'm not experienced with.

**Comment by wgranicki** [comment]
I think the problem is here
 https://github.com/bit-team/backintime/blob/29e874f5080f7cf85c8297ea807cb8070a427c10/qt/serviceHelper.py#L104

If I understand it correctly, only `nice` and `/usr/bin/nice` are accepted (whitelisted) as outputs of `which`, but in my case it's `/usr/sbin/nice`

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
> If I understand it correctly, only `nice` and `/usr/bin/nice` are accepted (whitelisted) as outputs of `which`, but in my case it's `/usr/sbin/nice`

Makes sense. `sbin` is for super-user only. Maybe Manjaro allow `nice` for `root` only?

I ask myself if we need `nice` at this point?

And I also ask if Manjaro offers a way to use `nice` in our case?

**Comment by wgranicki** [comment]
I don't think it's a Manjaro problem, because there's also `/usr/bin/nice` present. But I'm running BIT as root, because I want to backup the whole system, and not just the parts accessible to my user.

**Comment by wgranicki** [comment]
>Why does `backintime` appear in `/usr/sbin`? It definitely should not. How did you install it?

It appears both in `/usr/bin` and `/usr/sbin`. As I said before, I installed it from the official Manjaro repository via `pacman`

**Comment by emtiu** [comment]
Thanks for your reporting and testing. I can reproduce the error on a virtual Manjaro 21.3.7 installation.

It's a total mystery to me why the system tries to use `/usr/sbin/nice`, especially since `/usr/sbin` is just a symlink:
```
[manjaro@manjaro ~]$ ls -ld /usr/sbin
lrwxrwxrwx 1 root root 3  9. Jun 09:53 /usr/sbin -> bin
```
As far as I can see, this is different on Debian and Ubuntu, so we _are_ dealing with some sort of Manjaro specialty here.

A dirty workaround would be to include `/usr/sbin/nice` in the whitelist, but @brummer-simon has indicated that this might not even eliminate all problems:

> I have the same Issue and disabling "nice" doesn't help because the next the invalid command complains is "/usr/sbin/backintime". I'll resort to creating backups based on an interval. If you need move details on the issue feel free to contact me.

Also, I see the same confusing behavior as reported by @wgranicki:
> All of this seems quite strange, because neither `which` nor `whereis` report the `/usr/sbin/` locations - regardless if run as root or not. But somehow BIT finds those and throws exceptions.

So, the prize question is: **Why does backintime, _on Manjaro_, try to use binaries from `/usr/sbin` (such as `/usr/sbin/nice`) when run as root?**

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
My Debian doesn't have a `/usr/sbin/nice`.

Maybe Manjor `whereis` always sort the entries descending by there user rights.

In the reported case BiT run as root so the `sbin` folder is used instead.

**Comment by emtiu** [comment]
> Maybe Manjor `whereis` always sort the entries descending by there user rights.
> 
> In the reported case BiT run as root so the `sbin` folder is used instead.

It's not quite so simple, because on the _terminal_, `whereis nice` does _not_ point to a `/usr/sbin` location (and neither do other tools):
```
[manjaro ~]# whereis nice
nice: /usr/bin/nice /usr/include/nice [...]
[manjaro ~]# which nice
/usr/bin/nice
[manjaro ~]# type nice
nice is /usr/bin/nice
```

My best guess is that the reason is hidden in some place where Manjaro and Python meet (like default paths in the `sys` or `os` modules, or something).

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Did you tried to run `whereis nice` via `sudo` *and*  as `root` (sudo su)?

I see [this code](https://github.com/bit-team/backintime/blob/29e874f5080f7cf85c8297ea807cb8070a427c10/qt/serviceHelper.py#L110-L116) and would say bit does nothing less then just call `whereis nice` but as `root`.

**Comment by wgranicki** [comment]
I have tested it in all three variants: as regular user, with `sudo` and as root.
Every time the results are the same, and they never include the symlinked `sbin`, only `/usr/bin/nice`

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Good reason for me to create a Manjaro machine...

**Comment by wgranicki** [comment]
Just in case, I have also tested running `Popen(['which', 'nice'], stdout = PIPE).communicate()[0].strip().decode()` in Python console (again as regular user, with `sudo` and as root) and the results are the same as before, so it's definitely `pkexec`'s fault

**Comment by emtiu** [comment]
From a quick test, it looks like in Ubuntu (and probably Debian), `pkexec` works similarly – **but** `/usr/bin` and `/usr/sbin` are actually separate folders.

Therefore, `pkexec nice [program]` only suggests executables in `/usr/sbin` if they really exist there.

```
$ pkexec which nice
/usr/bin/nice
$ pkexec which fsck
/usr/sbin/fsck
```

This makes me think that it's a bad design decision by Manjaro to symlink `/usr/sbin` to `/usr/bin`, but there's nothing backintime can do about that.

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Maybe we can question why BIT needs `nice` and if it can't use `/usr/bin/nice` by default on all systems not matter if run as root or user. I don't understand the background here.

If not this should be reported to Manjaro.
Maybe they have a better idea.

**Comment by wgranicki** [comment]
> This makes me think that it's a bad design decision by Manjaro to symlink `/usr/sbin` to `/usr/bin`, but there's nothing backintime can do about that.

I don't think that's a bad decision in itself... But combined with this ↓ it creates the mess we see
```
$ pkexec env | grep PATH
PATH=/usr/sbin:/usr/bin:/sbin:/bin:/root/bin
```

**Comment by noyannus** [comment]
> > I have the same Issue and disabling "nice" doesn't help because the next the invalid command complains is "/usr/sbin/backintime". I'll resort to creating backups based on an interval. If you need move details on the issue feel free to contact me.
> 
> Same problem here

Still a problem. After disabling `nice` and `ionice` appears the "non-whitelisted command/parameter" `/usr/sbin/backintime` dialog.

backintime-qt 1.3.3-1  
backintime 1.3.3-1 
Both installed from Manjaro "Community" repo (can't remember which package manager).

Operating System: Manjaro Linux
KDE Plasma Version: 5.26.5
KDE Frameworks Version: 5.103.0
Qt Version: 5.15.8
Kernel Version: 6.1.12-1-MANJARO (64-bit)
Graphics Platform: X11

**Comment by buhtz** [comment]
Please open a ticket at Manjaro and/or Arch about it and leave us a link here. Because it seems to be Manjaro specific I would like to hear there expertise and opinion about it.

**Comment by noyannus** [comment]
[Manjaro post.](https://forum.manjaro.org/t/backintime-balks-at-udev-setting-since-2018/135579)

**Comment by yochananmarqos** [comment]
@buhtz Manjaro packager here. Please define how this issue is specific to Arch / Manjaro. I skimmed through this issue since it was brought to my attention on the [Manjaro forum](https://forum.manjaro.org/t/backintime-balks-at-udev-setting-since-2018/135579), but it's very hard to follow.

If there's something to be done packaging-wise, I'm happy to do it.

The Manjaro[ `backintime`](https://gitlab.manjaro.org/packages/community/backintime) split package can be found on our GitLab.

By the way, I follow [Fedora](https://src.fedoraproject.org/rpms/backintime/blob/rawhide/f/backintime.spec) more than I do the AUR package.

**Comment by emtiu** [comment]
> Please define how this issue is specific to Arch / Manjaro.

Thanks for coming in :) Here's what's happening:

1. Manjaro has `/usr/sbin` symlinked to `/usr/bin`.
2. Running the "BackInTime (root)" application uses `pkexec` to elevate permissions.
3. Under `pkexec`, paths are configured such that `which foo` returns `/usr/sbin/foo` _if it exists_. Else, it returns `/usr/bin/foo`.
4. backintime looks for the `nice` executable by calling `which nice`, **but** it's configured to only ever allow `/usr/bin/nice` as the answer. Else, the abovementioned Exception is thrown.

Now, if you run "BackInTime (root)" on Manjaro, the following things happen:
2 leads to 3, which leads to `which nice` returning `/usr/sbin/nice` because of 1, which leads to 4.

My best guess as to why this happened is that:
* `/usr/sbin/nice` is not, and has never been, a thing, which is why backintime has always gotten away with only accepting `/usr/bin/nice` as the return value of `which nice`.
* Manjaro and Arch (uniquely?) chose to symlink `/usr/sbin` to `/usr/bin`.
* Because of the way `pkexec` handles paths, this problems occurs only:
  * on Manjaro and Arch,
  * and only for "BackInTime (root)".

**Comment by yochananmarqos** [comment]
In summary, it's not a Manjaro issue then. As it seems you are aware, Manjaro inherits almost everything from Arch.

Why would the same thing not happen on Fedora, for example? Arch-based distros are not the only ones that don't use `/usr/sbin/`.

**Comment by emtiu** [comment]
Well, if nothing changes on the distro and packaging side of things, our best way to keep "BackInTime (root)" working on Arch and Manjaro is probably to do one of the things that @wgranicki suggested:

> I see at least two options here:
>     * whitelist `/usr/sbin/nice` (probably also `/bin/nice` and `/sbin/nice`), not just `/usr/bin/nice`
>     * modify the script to just check if `nice` is available on the system instead of verifying (whitelisting) the full path, or parse the output of `which` for `*bin/nice`

Any opinions on the best way forward?

**Comment by yochananmarqos** [comment]
Well, I can't do anything on the packaging end unless I patched upstream code. I'm not a programmer, so I won't be doing that unless there's a good PR that could resolve it.

**Comment by aryoda** [comment]
> * whitelist `/usr/sbin/nice` (probably also `/bin/nice` and `/sbin/nice`), not just `/usr/bin/nice`

I vote for implementing this option since the underlying code was introduced in 2017 (https://github.com/bit-team/backintime/commit/42b2e003ae43f89bccb369d92ce7f4b78f7ca520) to fix a highly critical CVE in the service helper (which runs with elevated rights) and allowed arbitrary arguments and is now too restrictive in above edge cases.

The package maintainers cannot fix this decently (without e.g. playing with the `path` env var when starting BiT which I'd consider really dirty) so i**t is to us to fix this.**

**Generally (out of scope here):** Our current dbus service helper is IMHO a security risk and modifying udev rules should be done without a permanently running daemon with elevated rights but by asking for elevated rights for `backintime-qt` only when the user wants to change the udev rules...

**Comment by buhtz** [comment]
Just for my learning: What does whitelisting technically mean in the context of this Issue?

**Comment by emtiu** [comment]
I'm not sure if it's a standard technique, but backintime does it like this: It is configured to run `which nice`, but only accepts `/usr/bin/nice` as an answer. If an attacker had modified paths so that `which nice` pointed to `/tmp/nice`, for example, then backintime would refuse to run `/tmp/nice`.

In this case, it is biting us in the ass, because through a chain of unfortunate circumstances, Arch and Manjaro create the illusion that `/usr/sbin/nice` exists, even though it doesn't really – and backintime refuses to run that, because it only accepts `/usr/bin/nice`.

The current best idea for a fix is to expand the whitelist, so that `/usr/sbin/nice` will also be accepted.

**Comment by buhtz** [comment]
How to reproduce the error? How to start BIT as root on Manjaro/Arch?

**Comment by wgranicki** [comment]
> How to reproduce the error? How to start BIT as root on Manjaro/Arch?

Well, in the Menu I have two entries: "Back In Time" (which links to `backintime-qt`) and "Back In Time (root)" (which links to `/usr/bin/backintime-qt_polkit %f`)

**Comment by buhtz** [comment]
Note to myself: I tried to contact a security expert via email about that issue and waiting for feedback.

**Comment by buhtz** [comment]
> > I see at least two options here:
> > * whitelist `/usr/sbin/nice` (probably also `/bin/nice` and `/sbin/nice`), not just `/usr/bin/nice`
> 
> Any opinions on the best way forward?

So I got feedback from the security expert at Suse who was involved in hardening this part of the code.
He do not see a security problem with this approach (whitelisting).

So we have decision: Whitelisting the missing path.

I won't takle it in the next days. Feel free to join if someone wants. And I am assuming I'll find some things to refactor. As easy as it sounds it will never be easy. ;)

**Comment by buhtz** [comment]
Sorry, guys for warming this up.
I am not able to reproduce the error message. Can you tell me how?

I installed the latest BIT version 1.6.0-dev on a Manjaro VM.
I run `backintime-qt_polkit` (which is the same as running "Back In time (root)" from applications menu). I setup a local profile with Udev scheduling.

I watched the terminal output and also "sudo journalctl -f". I got no error message like the initial one.

But I need to reproduce this with BIT itself to increase my understanding of the serviceHelper.py.

**EDIT**:

> 4. backintime looks for the nice executable by calling which nice, but it's configured to only ever allow /usr/bin/nice as the answer. Else, the abovementioned Exception is thrown.

I am not sure if this is correct. I see no whitelist in the code. The second parameter in `self._which('nice', '/usr/bin/nice')` seems to be a [fallback](https://github.com/bit-team/backintime/blob/74e2030e017f9126eb5b75c643c12413c11b52dc/qt/serviceHelper.py#L107-L119) not a allow-only-this-thing.

```py
        # ...
        self.nice = self._which('nice', '/usr/bin/nice')
        # ...


    def _which(self, exe, fallback):
        proc = Popen(['which', exe], stdout=PIPE)
        ret = proc.communicate()[0].strip().decode()
        if proc.returncode or not ret:
            return fallback

        return ret
```

**Comment by wgranicki** [comment]
I'm on version 1.5.4 (and Manjaro 25.0.2), and the problem persists. I don't see any changes to that part of the code between 1.5.4 and dev, so I assume my considerations below are valid.

> The second parameter in `self._which('nice', '/usr/bin/nice')` seems to be a [fallback](https://github.com/bit-team/backintime/blob/74e2030e017f9126eb5b75c643c12413c11b52dc/qt/serviceHelper.py#L107-L119)

If I understand it correctly (I'm not a Python programmer), `'/usr/bin/nice'` is a fallback in case `Popen(['which', exe], stdout=PIPE)` returns an error or an empty response. But this is not the case – it returns a response containing `'/usr/sbin/nice'`.

>I see no whitelist in the code.

If you look [a bit farther in the code](https://github.com/bit-team/backintime/blob/74e2030e017f9126eb5b75c643c12413c11b52dc/qt/serviceHelper.py#L131-L153), the result of `which` is sanitized, and it fails on comparison in line 151 (hence error message: "Parameter 'cmd' contains non-whitelisted cmd/parameter (/usr/sbin/nice)").

Since I don't have a test system (and really don't want to screw up my working one) and I'm not a Python programmer, I took the liberty to ask an AI-assistant for a suggestion on how to resolve the problem, and this is what it came up with:

_What You Should Do_
To fix this, you should normalize the paths to avoid false mismatches.

_Option 1: Compare Real Paths (Best)_
This ensures you match even if the path is a symlink:
```
elif os.path.realpath(parts[0]) != os.path.realpath(self.backintime):
    raise InvalidCmd("Parameter 'cmd' contains non-whitelisted "
                     f"cmd/parameter ({parts[0]})")
```

Same goes for all binaries in the whitelist:
```
while parts:
    for c, switches in whitelist:
        if os.path.realpath(parts[0]) == os.path.realpath(c):
            parts.pop(0)
            while parts and re.match(switches, parts[0]):
                parts.pop(0)
            break
    else:
        break
```

_Option 2: Canonicalize at Init Time_
When you assign the binary paths, resolve the real path immediately:

`self.nice = os.path.realpath(self._which('nice', '/usr/bin/nice'))
`

While I don't know if the above solutions are correct, they do look _nice_ ;)

**Comment by buhtz** [comment]
Thank you for your feedback and analysis. It was a bit late for me yesterday. :smile:
Maybe I can reproduce it when I try it the next time.

I can reproduce the difference between `pkexec which` and `sudo which`, of course. But I fail to trigger the error message with BIT itself. I'll try again.

**Comment by buhtz** [comment]
Can reproduce now. Not sure, what the difference was...

**Comment by buhtz** [comment]
There are `backintime` binaries in `/usr/sbin/` on Manjaro. Why? What is the design decision behind that?

**Comment by yochananmarqos** [comment]
@buhtz `/usr/sbin/` is a symlink. All binaries are in `/usr/bin/`.

**Comment by buhtz** [comment]
PR #2187 might solve the problem.
But I am not sure if I just miss to trigger the error or if it is really fixed.
